### ISSUE TITLE
How can ray register all related classes efficiently

### ISSUE THREAD

**Issue by raopku** [ray-project/ray#448]
for example:
To serialize a keras model, i follow the prompts of error logs to register class one by one.

```
ray.register_class(Sequential)
ray.register_class(Dense)
ray.register_class(Activation)
ray.register_class(Flatten)
from tensorflow.python.framework.ops import Tensor
from keras.engine.topology import Node
from tensorflow.python.ops.variables import Variable

ray.register_class(Tensor)
ray.register_class(Node)
ray.register_class(Variable)
```
However, after register these classes. It still does not work. 
`Exception: Ray does not know how to serialize objects of type <type 'function'>. To fix this, call 'ray.register_class' with this class.`
So I have to keep registering class...
How can ray register all related class efficiently?

**Comment by robertnishihara** [comment]
Good question, this is something that should be improved a lot.

**Some thoughts:** Let me know if this helps.

1. **Simple workaround.** As a workaround, assuming that pickle or cloudpickle can successfully serialize the Keras model, you could just work around Ray's serialization limitations by doing the serialization yourself. For example, to return a Keras model from a remote function, you could try doing something like this.

    ```python
    import cloudpickle

    # Create a remote function that uses a Keras model.
    @ray.remote
    def f():
      # Create a Keras model
      model = ...
      return cloudpickle.dumps(model)

    # Call the remote function to get the model and deserialize it by hand.
    model_id = f.remote()
    serialized_model = ray.get(model_id)
    model = cloudpickle.loads(serialized_model)
    ```

    Similarly, to pass a Keras model into a remote function, you could try something like the following.

    ```python
    import cloudpickle

    # Define a remote function that uses a Keras model.
    @ray.remote
    def f(serialized_model):
      # Deserialize the model.
      model = cloudpickle.loads(serialized_model)

    # Create a Keras model.
    model = ...
    # Serialize the Keras model.
    serialized_model = cloudpickle.dumps(model)

    # Pass the serialized model into the remote function.
    f.remote(serialized_model)
    ```

    Note that `pickle` and `cloudpickle` can't serialize *everything*, and TensorFlow models in particular don't always work. If the thing you're trying to pass in and out of remote functions is not serializable with `cloudpickle`, then the above approach won't work out of the box. That said, you could replace `cloudpickle.dumps` and `cloudpickle.loads` with your own functions if *you* know how to serialize the Keras model.

2. **Do you actually want to serialize the Keras model.** Constructing a TensorFlow model can be a slow process, so there may be much more efficient ways of doing what you're trying to do. For example,

    - Maybe what you want is to construct the neural net once inside of a Ray actor. That way you can just construct the model once inside of the actor constructor and never have to serialize/deserialize the model itself.
    - Maybe you can get away with just shipping the neural net weights (e.g, as a numpy array or a dictionary of numpy arrays). This will be *much* more efficient than serializing the actual model object.

3. Either way, we really should improve the `register_class` stuff. One option would be to provide a method like `ray.serialization.register_classes(object)`, which attempts to recursively register all of the relevant classes needed to serialize `object`.

4. In the error message you posted above,

    ```
    Exception: Ray does not know how to serialize objects of type <type 'function'>. To fix this, call 'ray.register_class' with this class.
    ```

    It looks like it's having trouble serializing a function. You might try fixing that by doing

    ```python
    f = lambda x: x + 1
    ray.register_class(type(f), pickle=True)
    ```

5. We should move `ray.register_class` to `ray.experimental.register_class` to indicate that this API will likely change as we improve things.

**Comment by raopku** [comment]
I know your way to serialize a keras model. It should be a feasible way to serialize keras models.
However i think it will be more efficient to serialize the necessary information which can be used to reconstruct a keras model.
for examle:
```
model = Sequential()
model.add(Flatten(input_shape=(1,) + (2,)))
model.add(Dense(16))
model.add(Activation('linear'))


config = {
    'class_name': model.__class__.__name__,
    'config': model.get_config(),
}

```

Then in the actor __init__ function:
`model = model_from_config(model_config, custom_objects={})`

The model is been serialized only once. When the agents are rolling out, only the weights will be shared between different agents.

**Comment by robertnishihara** [comment]
You can apply the approach in my previous comment to `model_config` instead of to `model`. That should work, right?

**Comment by robertnishihara** [comment]
This should be addressed by #550. Now `register_class` is removed from the API and should be called automatically under the hood.

Please reopen if you run into any problems.
